Add example dataset loading functionality

vetschn opened the pull request:
This still needs some thinking. Also we need to somehow provide proper setups, not just the input data.

As we have the same input data for multiple examples but different settings in the `config.toml` (e.g. carbon nanotube with either QTBM / phonons or coulomb Screening), we could maybe have all the small input text files directly in the repo as well and then just load the large raw data on demand. We could perhaps have symlinks to the input data that are dangling until the data is actually loaded. This way we also could ensure that the input files remain compatible with the code version.

The other option would be to have example archives including configs sitting in some other repo and loading specific subfolders. I have not found a way to download only selected subfolders on github, huggingface or zenodo, so probably won't work nicely.

- [ ] Gather all the data
- [ ] Decide how to best set this up
- [ ] Set up all the different examples and document them a little

vetschn committed:
Add example dataset loading functionality and manifest file.

This still needs some thinking. Also we need to somehow provide proper setups, not just the input data.

almaeder commented:
I believe this should also include the CLI to run the example and remove the need for the run scripts.
The cli would also allow us to fetch a certain example nicely.

almaeder commented:
"The other option would be to have example archives including configs sitting in some other repo and loading specific subfolders. "
Not a huge fan since the configurations will still be rapidly changing.

almaeder commented:
We could change the enforcement of a fixed file name.

almaeder committed:
Implements CLI functionality and example dataset management

- Added new dependency: typer.
- Included existing dependency: threadpoolctl
- Introduced CLI entrypoint in `src/quatrex/cli/main.py`.
- Updated `pyproject.toml` to include CLI script configuration.
- Improved example dataset loading
- Enhanced compute configuration with threading options.
- Communication configuration issue.

almaeder committed:
Add comprehensive tests for CLI commands and refactpr config parsing tests

almaeder committed:
Add GitHub Actions workflow; update environment, enable quatrex tests

- Enforce python 3.12 for slice hashing, adding test dependencies for quatrex
- Refactor environment-dev.yaml to update dependency specifications
- Fix docstring formatting and remove time_range in SCBA

almaeder commented:
Tests are expected to fail for quatrex since the dev branch is brocken.

alexnick83 commented on .github/workflows/tests.yaml:
```diff
@@ -17,7 +17,7 @@ jobs:
         uses: conda-incubator/setup-miniconda@v3
         with:
-          python-version: "3.11"
+          python-version: "3.12"
```
Let's move this to 3.13 since it is supported now on all dependencies. Going forward, and after an initial release, we should make a plan on how many Python versions we will support and for how long.

alexnick83 commented on .github/workflows/tests.yaml:
```diff
@@ -33,8 +33,8 @@ jobs:
       - name: Run MPI Tests
         run: |
-          NUMBA_DISABLE_JIT=1 mpiexec -np 6 pytest --only-mpi tests/qttools
+          NUMBA_DISABLE_JIT=1 mpiexec -np 6 pytest --only-mpi tests
```
Shall we resume testing with Numba? Or are we planning to have a separate set of tests with it?

vetschn replied:
This still tests all the numba kernels but just turns of the JIT which allows us to get a more accurate coverage rate (jitified kernels were counted as onlz one line)

alexnick83 replied:
I see. To be clear, according to the [documentation](https://numba.readthedocs.io/en/0.61.2/reference/envvars.html#envvar-NUMBA_DISABLE_JIT), setting the flag makes it so the original Python code is called. It may also make sense to run (just) the relevant tests a second time, with the JIT enabled, to ensure that the Numba-optimized code is actually running and returning the correct results.

alexnick83 commented on environment-dev.yaml:
```diff
@@ -3,74 +3,22 @@ channels:
-  - xz=5.4.6
-  - yaml=0.2.5
-  - zlib=1.2.13
+  - black>=24.8.0
```
Here, we should just provide an exact conda working environment, after we remove modules installed with pip and special flags. We can always provide general dependencies and min/max versions in the setup.

alexnick83 committed:
Added fallback for the case where the sparsity pattern is not given. Added static methods for loading the Hamiltonian in regular sparse format (COO) and extracting blocks.

alexnick83 committed:
Updated the automatic determination of the electron energy window to use the new static methods introduced in the last commit.

alexnick83 reviewed (changes requested):
The general structure is excellent. I pushed a couple of commits to enable the example to run (needs determination of the energy window, which ends up having 14k points), but these should be further refactored. We also need a smaller test that can fit in the runner.

vetschn commented on src/quatrex/cli/main.py:
```diff
@@ -0,0 +1,208 @@
+
+quatrex_cli = typer.Typer(
+    add_completion=False,
+    rich_markup_mode="markdown",
```
Here we'll need to also add `pretty_exceptions_show_locals=False` to avoid crazy long tracebacks

almaeder committed:
Python 3.13

almaeder committed:
Remove run_example, example config part of repo

almaeder committed:
Format tests

almaeder committed:
Enable NUMBA tests

almaeder committed:
Fixed negative indexing

awinka reviewed (approved):
LGTM

almaeder committed:
Dist example, only eigvalsh, example file cache

alexnick83 committed:
Amended spillover correction in bd_sandwich_distr to take into accout symmetries.

alexnick83 committed:
Split qttools and quatrex tests to separate CI steps.

alexnick83 committed:
Amended script for coverage with MPI.

alexnick83 committed:
Run test step even if previous one failed.

almaeder committed:
Combining of coverage reports

almaeder committed:
CLI rank 0 prints, corrupted file cache, enable numba in quatrex tests

almaeder committed:
Problematic sequential blas under OMP

almaeder committed:
Naive reference test

almaeder commented:
It is not yet so great how the files are handled with the manifest and the naming that I did.
